doc/flamegraphs.md

```diff
@@ -0,0 +1,108 @@
+## Flame Graphs
+
+Flame graphs are a way of visualizing software CPU and heap profiles. Any tool
+that can create stack data in the correct format can be used to generate a flame
+graph. You can learn more about flame graphs
+[here](http://www.brendangregg.com/flamegraphs.html).
+
+Example flame graph (click to zoom):
+
+![Example Flamegraph](/contrib/flamegraphs/example.svg)
+
+### Installing The Flame Graph Scripts
+
+The upstream GitHub repo for flame graphs is at
+[brendangregg/FlameGraph](https://github.com/brendangregg/FlameGraph). To run
+the scripts you'll need a Perl installation on your system.
+
+There is a small patch in the Bitcoin repository at
+[contrib/flamegraphs/bitcoin-annotations.patch](/contrib/flamegraphs/bitcoin-annotations.patch)
+that you can apply to generate flame graphs that are color coded for Bitcoin.
+You can apply it using a command like this:
+
+```bash
+# Assuming you're at the top directory in the FlameGraph repo
+$ patch -p1 <path/to/bitcoin-annotations.patch
+```
+
+### Installing Perf
+
+In general you will need a version of the
+[perf](https://perf.wiki.kernel.org/index.php/Tutorial) command that matches the
+kernel that you are using. Perf is maintained along with the kernel source tree,
+and is available on most Linux distributions:
+
+```bash
+# Install perf on Debian.
+sudo apt-get install linux-perf
+
+# Install perf on Fedora.
+sudo dnf install perf
+```
+
+### Generating Flame Graphs On Linux Using Perf
+
+To generate flame graphs with labeled function names you'll need a build of
+Bitcoin that has debugging symbols enabled (this should be the case by default).
+You can check that your executable has debugging symbols by checking for a
+`.debug_str` section in the ELF file:, e.g.
+
+```bash
+# If this produces output your binary has debugging symbols.
+$ readelf -S src/bitcoind | grep .debug_str
+```
+
+If you want kernel annotations (optional) then you should also make sure that
+the `kernel.perf_event_paranoid` sysctl option is set to -1:
+
+```bash
+# Optional, enable kernel annotations
+sudo sysctl kernel.perf_event_paranoid=-1
+```
+
+Now you're ready to generate a profile. To profile a running bitcoind process
+for 60 seconds you can use the following command:
+
+```bash
+# Profile bitcoind for 60 seconds
+perf record -g --call-graph dwarf -F 101 -p $(pidof bitcoind) -- sleep 60
```

Comment:
`pidof` on macOS can be installed with `brew install pidof`. `perf` is `gperf` on macOS, although it doesn't recognize `--call-graph`.